### DISCUSSION TITLE
sample JSON  workflow with  memory blocks defined  and referenced

### DISCUSSION THREAD

**Discussion by mremadevi** [elsa-workflows/elsa-core#5296] [Q&A]
Please provide sample JSON  workflow with  memory blocks defined  and referenced

**Comment by glime-ai** [comment]
Certainly! Below is a sample JSON workflow that includes memory blocks defined and referenced. This example is based on the information provided in the context, specifically from the "Loading workflows from JSON" guide and the "Programmatic workflows vs designer workflows" section.

```json
{
  "id": "SampleWorkflow-v1",
  "definitionId": "SampleWorkflow",
  "name": "Sample Workflow",
  "version": 1,
  "isPublished": true,
  "isLatest": true,
  "root": {
    "id": "Flowchart1",
    "type": "Elsa.Flowchart",
    "activities": [
      {
        "id": "WriteLine1",
        "type": "Elsa.WriteLine",
        "text": {
          "typeName": "String",
          "expression": {
            "type": "Literal",
            "value": "Hello World!"
          },
          "memoryReference": {
            "id": "WriteLine1:input-1"
          }
        }
      },
      {
        "id": "ReadLine1",
        "type": "Elsa.ReadLine",
        "result": {
          "typeName": "String",
          "memoryReference": {
            "id": "UserInput"
          }
        }
      },
      {
        "id": "WriteLine2",
        "type": "Elsa.WriteLine",
        "text": {
          "typeName": "String",
          "expression": {
            "type": "JavaScript",
            "value": "`You entered: ${UserInput}`"
          },
          "memoryReference": {
            "id": "WriteLine2:input-1"
          }
        }
      }
    ],
    "connections": [
      {
        "source": "WriteLine1",
        "target": "ReadLine1",
        "sourcePort": "Done",
        "targetPort": "In"
      },
      {
        "source": "ReadLine1",
        "target": "WriteLine2",
        "sourcePort": "Done",
        "targetPort": "In"
      }
    ]
  }
}
```

In this workflow:

- `WriteLine1` is an activity that writes "Hello World!" to the console.
- `ReadLine1` is an activity that reads user input and stores it in a memory block with the ID `UserInput`.
- `WriteLine2` is an activity that writes the user input back to the console, using the memory block `UserInput` to retrieve the value.

The `memoryReference` property is used to define and reference memory blocks within the workflow. In this example, `WriteLine1:input-1` and `WriteLine2:input-1` are memory blocks that are defined and referenced by the `WriteLine` activities to store and retrieve the text to be written to the console.

Please note that this is a simplified example and actual workflows may include more complex structures and activities.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by mremadevi** [reply]
Variable 'UserInput' is not defined as workflow variable in given sample. Isn`t that required?

The above sample gives error UserInput is not defined

**Reply by mremadevi** [reply]
I am using Elsa studio 3.2 , and Json generated doesn't have a memory reference block  with expressions . The UI doesn`t have an option to bind the variable either . How is that getting added


![image](https://github.com/elsa-workflows/elsa-core/assets/167574820/d6ae79e4-4fe8-4e04-bb1a-a7f6e3694ac8)


![image](https://github.com/elsa-workflows/elsa-core/assets/167574820/d8f0f065-2bd7-46e9-9950-3ec6a56cb818)
